CRYPTO.md

```diff
@@ -0,0 +1,146 @@
+
+|        Key Store Alias Element | Size (Bytes) |
+|-------------------------------:|:-------------|
+|          Application Parameter | 32           |
+|              Key Handle Length | 1            |
+|                     Key Handle | 0-255        |
+
+This means that we can look up the signing key directly from an offset of the
+contents of the authentication message payload. Notice the overlap:
+
+| U2F AUTHENTICATE Request Field | Size (Bytes) |
+|-------------------------------:|:-------------|
+|            Challenge Parameter | 32           |
+|          Application Parameter | 32           |
+|              Key Handle Length | 1            |
+|                     Key Handle | 0-255        |
+
+This would make it impossible to use a signing key on anything but the same
+application parameter and key handle.
+
+This implementation has one big positive: all signing happens inside the
+hardware, significantly limiting possible key exposure. This comes at the cost
+of a new key in the key store for each registration. The maximum number of keys
+in the hardware is unknown (to me).
+
+One might also suggest that a user could, at confirmation for key creation,
+enter metadata to uniquely identify the key and then we could have support for
+on-device revocation. However, this may not be desirable. Attaching metadata to
+keys could leak privacy information. Further, device-side revocation still
+doesn't revoke the key on the server. Thus the server-side would still be
+vulnerable.
+
+On the other hand, one bit of metadata which might be nice to attach at key
+creation time would be user confirmation and key security policy. This can be
+implemented without per-registration hardware keys. However, with them, the
+policy can be evaluated inside the key store code itself. For example, at key
+creation time the use could say "I want fingerprint verifciation for this key."
+Then, we can turn on the key store's fingerprint verification flag. This is a
+benefit of this implementation design. Otherwise, we have to do all of this
+policy evaluation in software (and possibly with a compromise on, say, a rooted
+phone).
+
+## Trusted Executable Environment
+
+It might be possible to get the best of both worlds by writing a [TEE
+process][tee]. However, this would come at the cost of significant additional
+complexity. The key store is already implemented as a TEE and might provide us
+this functionality in the future.
+
+# Conclusion
+
+FreeU2F is currently implemented with the Per-Registration Hardware Keys model.
+This was chosen for its ease of implementation and good security benefits.
+However, this might change in the future.
+
+# Remaining Issues
+## Counter Management
+
+We need a counter for the authentication reply. This is not currently
+implemented (we always return 0). Where should we store this?  Should this be
+per-registration or global? How should we protect it from outside modification?
```

Comment:
I guess also, how long are is the counter expected to be in use for, as that will presumably change the width we allow for the storage?